File: VB/Northwind.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Namespace Northwind

	<Persistent("Categories")> _
	Public Class Category
		Inherits XPLiteObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		<Key(True), Persistent("CategoryID")> _
		Private fCategoryID As Integer = -1

		<PersistentAlias("fCategoryID")> _
		Public ReadOnly Property CategoryID() As Integer
			Get
				Return fCategoryID
			End Get
		End Property
		Private fCategoryName As String
		Public Property CategoryName() As String
			Get
				Return fCategoryName
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)("CategoryName", fCategoryName, value)
			End Set
		End Property
		Private fDescription As String
		Public Property Description() As String
			Get
				Return fDescription
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)("Description", fDescription, value)
			End Set
		End Property

		<Association("ProductCategories")> _
		Public ReadOnly Property Products() As XPCollection(Of Product)
			Get
				Return GetCollection(Of Product)("Products")
			End Get
		End Property

		' a filtered collection
		Public ReadOnly Property AvailableProducts() As XPCollection(Of Product)
			Get
				Dim filter As CriteriaOperator = New OperandProperty("Discontinued") <> New OperandValue(True)
				Return New XPCollection(Of Product)(Me.Products, filter)
			End Get
		End Property
	End Class

	<Persistent("Products")> _
	Public Class Product
		Inherits XPLiteObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		<Key(True), Persistent("ProductID")> _
		Private fProductID As Integer = -1

		<PersistentAlias("fProductID")> _
		Public ReadOnly Property ProductID() As Integer
			Get
				Return fProductID
			End Get
		End Property

		Private fProductName As String
		Public Property ProductName() As String
			Get
				Return fProductName
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)("ProductName", fProductName, value)
			End Set
		End Property

		Private fCategory As Category
		<Persistent("CategoryID"), Association("ProductCategories")> _
		Public Property Category() As Category
			Get
				Return fCategory
			End Get
			Set(ByVal value As Category)
				SetPropertyValue(Of Category)("Category", fCategory, value)
			End Set
		End Property

		Private fUnitPrice As Decimal
		Public Property UnitPrice() As Decimal
			Get
				Return fUnitPrice
			End Get
			Set(ByVal value As Decimal)
				SetPropertyValue(Of Decimal)("UnitPrice", fUnitPrice, value)
			End Set
		End Property
		Private fDiscontinued As Boolean
		Public Property Discontinued() As Boolean
			Get
				Return fDiscontinued
			End Get
			Set(ByVal value As Boolean)
				SetPropertyValue(Of Boolean)("Discontinued", fDiscontinued, value)
			End Set
		End Property
	End Class

End Namespace

File: CS/Northwind.cs
using System;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
namespace Northwind {

    [Persistent("Categories")]
    public class Category : XPLiteObject {
        public Category(Session session) : base(session) { }

        [Key(true), Persistent("CategoryID")]
        int fCategoryID = -1;
        
        [PersistentAlias("fCategoryID")]
        public int CategoryID {
            get { return fCategoryID; }
        }
        string fCategoryName;
        public string CategoryName {
            get { return fCategoryName; }
            set { SetPropertyValue<string>("CategoryName", ref fCategoryName, value); }
        }
        string fDescription;
        public string Description {
            get { return fDescription; }
            set { SetPropertyValue<string>("Description", ref fDescription, value); }
        }

        [Association("ProductCategories")]
        public XPCollection<Product> Products {
            get { return GetCollection<Product>("Products"); }
        }

        // a filtered collection
        public XPCollection<Product> AvailableProducts {
            get {
                CriteriaOperator filter = new OperandProperty("Discontinued") != new OperandValue(true);
                return new XPCollection<Product>(this.Products, filter); 
            }
        }
    }

    [Persistent("Products")]
    public class Product : XPLiteObject {
        public Product(Session session) : base(session) { }

        [Key(true), Persistent("ProductID")]
        int fProductID = -1;
        
        [PersistentAlias("fProductID")]
        public int ProductID {
            get { return fProductID; }
        }

        string fProductName;
        public string ProductName {
            get { return fProductName; }
            set { SetPropertyValue<string>("ProductName", ref fProductName, value); }
        }
        
        Category fCategory;
        [Persistent("CategoryID"), Association("ProductCategories")]
        public Category Category {
            get { return fCategory; }
            set { SetPropertyValue<Category>("Category", ref fCategory, value); }
        }

        decimal fUnitPrice;
        public decimal UnitPrice {
            get { return fUnitPrice; }
            set { SetPropertyValue<decimal>("UnitPrice", ref fUnitPrice, value); }
        }
        bool fDiscontinued;
        public bool Discontinued {
            get { return fDiscontinued; }
            set { SetPropertyValue<bool>("Discontinued", ref fDiscontinued, value); }
        }
    }

}
